exprel/graph_extractor/extract.py

```diff
@@ -0,0 +1,331 @@
+                    yield g
+
+        if graph_type == "ud":
+            self.init_nlp()
+            for sen in tqdm(iterable):
+                doc = self.nlp(sen)
+                G, _ = ud_to_graph(doc.sentences[0])
+                yield G
+
+
+class FeatureEvaluator:
+    def __init__(self, graph_format="ud"):
+        self.graph_format = graph_format
+
+    def match_features(self, dataset, features):
+        graphs = dataset.graph.tolist()
+
+        matches = []
+        predicted = []
+
+        matcher = GraphFormulaMatcher(features, converter=default_pn_to_graph)
+
+        for i, g in tqdm(enumerate(graphs)):
+            feats = matcher.match(g)
+            for key, feature in feats:
+                matches.append(features[feature])
+                predicted.append(key)
+                break
+            else:
+                matches.append("")
+                predicted.append("")
+
+        d = {
+            "Sentence": dataset.text.tolist(),
+            "Predicted label": predicted,
+            "Matched rule": matches,
+        }
+        df = pd.DataFrame(d)
+        return df
+
+    def one_versus_rest(self, df, entity):
+        mapper = {entity: 1}
+
+        one_versus_rest_df = df.copy()
+        one_versus_rest_df["one_versus_rest"] = [
+            mapper[item] if item in mapper else 0 for item in df.label
+        ]
+
+        return one_versus_rest_df
+
+    def train_feature(self, cl, feature, data, graph_format="fourlang"):
+        feature_graph = default_pn_to_graph(feature)[0]
+
+        graphs = data.graph.tolist()
+        labels = self.one_versus_rest(data, cl).one_versus_rest.tolist()
+        path = "trained_features.tsv"
+        trained_features = []
+        with open(path, "w+") as f:
+            for i, g in enumerate(graphs):
+                matcher = DiGraphMatcher(
+                    g,
+                    feature_graph,
+                    node_match=GraphFormulaMatcher.node_matcher,
+                    edge_match=GraphFormulaMatcher.edge_matcher,
+                )
+                if matcher.subgraph_is_isomorphic():
+                    for iso_pairs in matcher.subgraph_isomorphisms_iter():
+                        nodes = []
+                        for k in iso_pairs:
+                            if feature_graph.nodes[iso_pairs[k]]["name"] == ".*":
+                                nodes.append(g.nodes[k]["name"])
+                        nodes_str = ",".join(nodes)
+                        label = labels[i]
+                        sentence = data.iloc[i].text
+                        f.write(f"{feature}\t{nodes_str}\t{sentence}\t{label}\n")
+                        trained_features.append(
+                            (feature, nodes_str, sentence, str(label))
+                        )
+
+        return self.cluster_feature(trained_features)[1]
+
+    def cluster_feature(self, trained_features):
+        def to_dot(graph, feature):
```

Comment:
eventually we should move to a proper programmatic generation of graphs, e.g. with [nx_pydot](https://networkx.org/documentation/stable/reference/generated/networkx.drawing.nx_pydot.to_pydot.html), or with something else entirely

Reply:
Yes but with custom functions we can fully customize the generation, e.g. mark nodes with colors. But I agree that we should search for a solution